sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/BatchGroupUtils.scala

```diff
@@ -0,0 +1,288 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.rapids.execution.python
+
+import scala.collection.mutable
+
+import ai.rapids.cudf
+import com.nvidia.spark.rapids._
+import com.nvidia.spark.rapids.RapidsPluginImplicits._
+
+import org.apache.spark.TaskContext
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.execution.SparkPlan
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+
+/**
+ * Basic functionality to deal with groups in a batch.
+ *
+ * It plays the similar role to the Spark 'PandasGroupUtils', but dealing with 'ColumnarBatch',
+ * instead of 'InternalRow'.
+ */
+private[python] object BatchGroupUtils extends Arm {
+
+  /**
+   * Returns the deduplicated attributes of the Spark plan, and the argument offsets of the
+   * keys and values.
+   *
+   * The deduplicated attributes are needed because the Spark plan may contain an attribute
+   * twice; once in the key and once in the value. For any such attribute we need to
+   * deduplicate.
+   *
+   * The argument offsets are used to distinguish grouping attributes and data attributes
+   * as following:
+   *
+   * argOffsets[0] is the length of the argOffsets array
+   * argOffsets[1] is the length of grouping attribute
+   * argOffsets[2 .. argOffsets[1]+2] is the argument offsets for grouping attributes
+   * argOffsets[argOffsets[1]+2 .. ] is the argument offsets for data attributes
+   */
+  def resolveArgOffsets(
+      child: SparkPlan,
+      groupingAttrs: Seq[Attribute]): (Seq[Attribute], Array[Int], Seq[Int]) = {
+
+    val dataAttrs = child.output.drop(groupingAttrs.length)
```

Comment:
Actually I had the same question  when investigating the CPU code before. 

I did some verification on normal Spark (not on DB yet), and seems the grouping attributes are always prepended to the original attributes in the child output.

Reply:
My concern here is like with the sorting assumption. We have a requirement for an API that is not documented, it is only assumed. If the API was obviously coupled with what uses it, then it would be less of an issue, but because it appears to be a generic API we need to document the assumptions about how it should be used.

Reply:
Got it, really good suggestion.

Added it as 
```scala
   * Besides, this API supposes the grouping attributes are placed before the others in
   * the output attributes of the plan.
```